src/lib/operators/abstract_aggregate_operator.hpp

```diff
@@ -0,0 +1,119 @@
+ */
+struct AggregateColumnDefinition final {
+  AggregateColumnDefinition(const std::optional<ColumnID>& column, const AggregateFunction function)
+      : column(column), function(function) {}
+
+  std::optional<ColumnID> column;
+  AggregateFunction function;
+};
+
+/*
+The AggregateFunctionBuilder is used to create the lambda function that will be used by
+the AggregateVisitor. It is a separate class because methods cannot be partially specialized.
+Therefore, we partially specialize the whole class and define the get_aggregate_function anew every time.
+*/
+template <typename ColumnDataType, typename AggregateType, AggregateFunction function>
+struct AggregateFunctionBuilder {
+  void get_aggregate_function() { Fail("Invalid aggregate function"); }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Min> {
+  auto get_aggregate_function() {
+    return [](const ColumnDataType& new_value, std::optional<AggregateType>& current_aggregate) {
+      if (!current_aggregate || value_smaller(new_value, *current_aggregate)) {
+        // New minimum found
+        current_aggregate = new_value;
+      }
+      return *current_aggregate;
+    };
+  }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Max> {
+  auto get_aggregate_function() {
+    return [](const ColumnDataType& new_value, std::optional<AggregateType>& current_aggregate) {
+      if (!current_aggregate || value_greater(new_value, *current_aggregate)) {
+        // New maximum found
+        current_aggregate = new_value;
+      }
+      return *current_aggregate;
+    };
+  }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Sum> {
+  auto get_aggregate_function() {
+    return [](const ColumnDataType& new_value, std::optional<AggregateType>& current_aggregate) {
+      // add new value to sum
+      if (current_aggregate) {
+        *current_aggregate += new_value;
+      } else {
+        current_aggregate = new_value;
+      }
+    };
+  }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Avg> {
+  auto get_aggregate_function() {
+    // We reuse Sum here and use it together with aggregate_count to calculate the average
+    return AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Sum>{}.get_aggregate_function();
+  }
+};
+
+template <typename ColumnDataType, typename AggregateType>
+struct AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Count> {
+  auto get_aggregate_function() {
+    return [](const ColumnDataType&, std::optional<AggregateType>& current_aggregate) { return std::nullopt; };
```

Comment:
Just like the other builders, this is extracted from the hash aggregate, so the following is mostly speculation:

In both hash and sort aggregate, we increase the counts directly while iterating over the column/segments.
Incrementing an int in the main loop might be more efficient than doing it in this function (no idea if there's a difference), but I believe there is an another reason:
For calculating the average, we do need both the sum, and the number of elements. The function for average is already used to calculate the sum, so we need to keep track of the number of elements manually. As we do this manually anyway, there is no reason to do it a second time in this builder.

Now you might argue that one could use const expressions to avoid calling this function at all, so we can really get rid of it. This would probably improve the performance, but handling many special cases can lead to worse readability. Any comment on this from @Bensk1 or @mrks ?

Reply:
Feel free to test if it makes a difference, but I'd keep it out of this PR. I agree on the readability being better, so the improvement should be measurable.